### ISSUE TITLE
[RFC] Plugin system for custom rules

### ISSUE THREAD

**Issue by marcelofabri** [realm/SwiftLint#2529]
## Motivation

We currently support custom rules as regexes, but often that's not the right tool for the job. Rules implemented with regexes on SwiftLint are usually the slower ones and also the ones with more false positives. Having first-class support to a plugin system would be essential for empowering users to automatically enforce specific patterns that don't necessarily belong to SwiftLint itself.

## Proposal

This proposal describes a flexible plugin system that doesn't require new rules to be written using Swift. We should provide tools to make it possible though, but that will be left as a future enhancement.

A plugin is an executable that will be run by SwiftLint and should respond to certain commands. Communication will happen via stdin and stdout using JSON unless specified.

A plugin maps 1:1 to a rule, so you can't implement more than one rule in the same plugin.

**Note**: so far this is just a proposal, no concrete implementation has started as far as I know.

### Commands

`plugin` represents the path to a generic custom plugin executable in the commands below.

#### `plugin rule_description`

This will return a JSON containing all the required data to create a `RuleDescription`. Will be run by SwiftLint when parsing the configuration and resolving the rules.

**Output:**

```json
{
    "identifier": "my_cool_rule",
    "name": "Cool Rule",
    "kind": "style",
    "non_triggering_examples": [
        "let x = 10"
    ],
    "triggering_examples": [
        "let x = 20"
    ],
    "min_swift_version": "4.1.0"
}
```

#### `plugin lint`

This will receive a JSON as input and should return another JSON containing the violations.

**Input**:
```json
{
    "structure": ... ,
    "syntax_map": ...,
    "path": "file.swift",
    "contents": "",
    "configuration": {
        "cool_number": 10
    }
}
```

Note: `path` is optional. I have mixed feeling whether `contents` should be populated if `path` is present.

`configuration` contains the raw configuration loaded from the configuration YAML file as described below in "Configuring a plugin".

**Output**:
```json
[
    {
        "severity": "warning",
        "reason": "Failed",
        "location": {
            "line": 1,
            "character: 10
        },
        "byte_offset": 203,
        "character_offset": 300
    }
]
```

Note: `location`, `byte_offset` and `character_offset` are exclusive (only one should be provided). `reason` is optional.

#### Alternatives considered

Another option is to have the `lint` command run only once and keep feeding it JSONs. This would have the benefit of being faster (as fixed costs such as loading dynamic libraries would be paid only once) but adds more complexity.

However, simplifying this might lead to unresolvable performance problems in the future, so I'd love more input on this topic.

### Enabling plugins in the configuration

On `.swiftlint.yml` there'll be a new `plugins` key:

```yml
plugins:
    - path/to/plugin
```

### Configuring a plugin

Configuring a plugin will be done in the same way as a configuration is provided to a rule, by adding a key on `.swiftlint.yml`:

```yml
my_cool_rule:
    cool_number: 10
```

All the values inside the key corresponding to the rule identifier will be provided as part of the JSON input as described in the `plugin lint` section.

### Caching

Caching should work without major changes. The only potential issue is that SwiftLint wouldn't be able to see updates on the binary implementation, forcing the user to use `--no-cache` while implementing the custom rule. This could be mitigated by having a `plugin version` command (described below in "Future Enhancements").

### Analyzer

Implement an `analyze` rule will be unsupported by now as it's still experimental.

### `swiftlint rules`

SwiftLint should list rules implemented in plugins in `swiftlint rules`.

However, it's not clear for me what would be the best option to expose the available configuration options. 

The easiest way would be adding another command in the plugin to return a String (similar to `configurationDescription` in the `Rule` protocol).

### `plugin autocorrect`

Autocorrect support is nice but it'll be left as a future enhancement since it should be an additive change. It should work in a similar way as `plugin lint`.

### `plugin version`

This should output to stdout a string containing the version (e.g. `0.1.0`) and exit the program. This will be called on the start of the lint/autocorrect command on SwiftLint to gather all plugin versions and use that to invalidate the violations cache if needed.

### `syntaxMap` and `structure` only when required

Right now, we'll ask SourceKit for the `structure` and `syntaxMap` for every file that needs to be linted using a plugin rule. However, it's common that a rule only need one or another (or even neither!). We could provide a way for the plugin to explicitly say what piece of information it needs and then SwiftLint would only provide that, speeding up the process.

### Docs generation for plugin rules

`plugin rule_description` can return non-triggering and triggering examples that could be used to generate something like Rules.md for the specific plugin.

### Opt-in rules

As described, all rules provided by plugins are enabled by default. We could add support to opt-in rules in plugins in the future (although I'm not quite sure what would be the real use case for it).

**Comment by marcelofabri** [comment]
@jpsim would love to hear your ideas about this!

**Comment by marcelofabri** [comment]
I made some progress on a proof of concept today: https://github.com/realm/SwiftLint/tree/mf-plugins.

However, it's pretty slow even with an empty rule:

### Top 10 slowest rules to lint on SwiftLint codebase
```
0,256: closure_end_indentation
0,301: opening_brace
0,364: colon
0,402: identifier_name
0,468: mark
0,480: operator_usage_whitespace
0,511: comma
1,879: identical_operands
2,842: let_var_whitespace
7,108: example
```

Sample plugin:

```swift
#!/usr/bin/swift

import Foundation

struct RuleDescription: Equatable, Codable {
    let identifier: String
    let name: String
    let description: String
    let kind: String
    let nonTriggeringExamples: [String]
    let triggeringExamples: [String]
    let corrections: [String: String]
    let deprecatedAliases: Set<String>
    let minSwiftVersion: String
    let requiresFileOnDisk: Bool
}

func lint() throws {
	let inputData = FileHandle.standardInput.readDataToEndOfFile()
	if inputData.isEmpty {
		exit(1)
	}
	let json = [
		"reason": "Custom reason",
		"severity": "warning",
		"character_offset": 2
	] as [String: Any]
	let data = try JSONSerialization.data(withJSONObject: [])
	puts(String(decoding: data, as: UTF8.self))
}

func ruleDescription() throws {
    let description = RuleDescription(identifier: "example", name: "Example",
                                      description: "Test", kind: "style", nonTriggeringExamples: [],
                                      triggeringExamples: [], corrections: [:],
                                      deprecatedAliases: [], minSwiftVersion: "4.0.0",
                                      requiresFileOnDisk: false)
    let encoder = JSONEncoder()
    encoder.keyEncodingStrategy = .convertToSnakeCase

    let json = try String(decoding: encoder.encode(description), as: UTF8.self)
    puts(json)
}

if CommandLine.argc < 2 {
    exit(1)
}

let command = CommandLine.arguments[1]
switch command {
case "lint":
    try lint()
case "rule_description":
    try ruleDescription()
default:
    exit(1)
}
```

I haven't profiled it yet, but I start to think that launching a process to lint each file is not a good idea 😅

**Comment by marcelofabri** [comment]
I was able to make it faster than some existing rules by using an UNIX socket instead of launching several processes:

```
0,446: mark
0,452: comma
0,461: operator_usage_whitespace
1,524: example <<<<<<<<<
1,887: identical_operands
2,693: let_var_whitespace
```

It's still not as fast as it could be because I'm not reusing the sockets from SwiftLint's side, so a new connection is created every time a file is linted. 

I haven't yet implemented actually spawning the server socket before linting starts, so there'd be some overhead there as well.

**Comment by marcelofabri** [comment]
Turns out at least 50% of the time is spent serializing to JSON:

<img width="567" alt="captura de tela 2018-12-27 as 02 04 10" src="https://user-images.githubusercontent.com/833072/50476117-c7400600-097b-11e9-93fd-5354438de436.png">

At this point, I wonder if it's really better to send `structure` and `syntax_map` to plugins versus letting them call SourceKit directly.

This is what I think we should do:

1. Allow the plugin to specify which information they need (`structure`, `syntax_map`, both or none) and make the default "none". Only plugins implemented in other languages or outside the "default" plugin setup would set that property (probably as part of `rule_description` - which at this point should be renamed to `plugin_description`.

2. Have a plugin template/dependency which should abstract that away calling SourceKit (as it's done currently when working on a normal rule) and dealing with sockets or serialization. That setup would require users to write rules in Swift, which I don't think it's a problem given that most people using a linter for Swift are probably comfortable with the language.

I don't expect many people to use (1) but given how easy it's to support, I think we should. The downsides that I can think of:

- support cost in the future
- having to document the plugin architecture, which we should do either way
- making the API format "public" as in "we shouldn't break it". However, I think this would be the case anyway as we wouldn't want to break existing plugins.

**Comment by Dschee** [comment]
After reading your comments I have a few questions:

1. Should the new plugin system replace custom rules based on regexes as they are currently supported?
2. Having a look at your example plugin, it looks to me like we could remove all controversial or rarely used opt-in rules and make them plugins instead, changing the purpose of SwiftLint more to a community standard and only accepting rules which are considered good practices. This way we could much simplify the complexity of SwiftLint and concentrate on the actual "good practices". This could also bring the entire community forward, while we could support a few different "flavors" officially, which would basically be different default SwiftLint configurations.
3. Goint into the direction of point 2 (even if we don't go it all the way) we should definitely support not only single plugins, but entire directories or plugin collections. For example, this could be a `.swiftlint.yml` entry:

```yaml
plugins:
    - path/to/custom/plugins/directory/*
    - https://github.com/Moya/SwiftLintPlugins
```

The direction this idea would take enables:
- Companies to enforce their custom plugins across all of their Swift projects (by providing a custom plugin collection maintained separately from the current project)
- Open Source Projects like Moya, Realm or Eureka to write plugins to enforce Best Practices for their specific frameworks

I'm not sure if this direction is in line with what you were imagining for this to be. These ideas just came into my mind and the initial implementation maybe should consider it so adding those features is easy.

**Comment by marcelofabri** [comment]
> Should the new plugin system replace custom rules based on regexes as they are currently supported?

No, but the new plugin system would be more powerful and preferred in many cases.

> Having a look at your example plugin, it looks to me like we could remove all controversial or rarely used opt-in rules and make them plugins instead

I think that's possible. Not sure about the existing rules, but new ones can follow this pattern.

> Goint into the direction of point 2 (even if we don't go it all the way) we should definitely support not only single plugins, but entire directories or plugin collections

Supporting remote plugins is trickier, but I agree that we should support multiple rules

**Comment by Serchinastico** [comment]
As a SwiftLint user that was about to fork the project to create my own rules, I think point 2 is a great approach.

It reminds me of AndroidLint (for Java/Kotlin), where you can add custom rules by creating a (gradle) module and extending some key library classes. Most people in their projects create their own set of rules and add them to their project's repository. I think using Swift is not a drawback at all and having a dependency dealing with the internals of processes communication smooths the learning curve.
